src/schema/v2/filterArtworksConnection.ts

```diff
@@ -568,47 +566,65 @@ const filterArtworksConnectionTypeFactory = (
       return { hits: null, aggregations: null, gravityOptions }
     }
 
-    return loader(gravityOptions).then(({ aggregations, hits }) => {
-      if (!aggregations || !aggregations.total) {
-        throw new Error(
-          "Expected filter results to contain a `total` aggregation"
-        )
-      }
-
-      const totalPages = computeTotalPages(
-        aggregations.total.value,
-        gravityOptions.size
+    const { aggregations, hits } = await loader(gravityOptions)
+    if (!aggregations || !aggregations.total) {
+      throw new Error(
+        "Expected filter results to contain a `total` aggregation"
       )
+    }
 
-      const connection = connectionFromArraySlice(
-        hits,
-        { first, last, after, before },
-        {
-          arrayLength: Math.min(
-            aggregations.total.value,
-            totalPages * gravityOptions.size
-          ),
-          sliceStart: gravityOptions.offset,
-        }
-      )
+    const hasRequestedCollectorSignals = isFieldRequested(
+      "edges.node.collectorSignals",
+      info
+    )
 
-      connection.pageInfo.endCursor = pageToCursor(
-        gravityOptions.page + 1,
-        gravityOptions.size
-      )
+    const artworks = await Promise.all(
+      hasRequestedCollectorSignals
+        ? hits.map((artwork) => {
+            return collectorSignalsLoader(artwork, ctx).then(
+              (collectorSignals) => {
+                artwork.collectorSignals = collectorSignals
+
+                return artwork
+              }
+            )
+          })
+        : hits
```

Comment:
I _think_ this will error out if `hasRequestedCollectorSignals` is false, since `hits` isn't an array of promises at that point.

(Otherwise all good re yr comment / this approach)

Reply:
I wondered that too because it's awkward to work with, especially before switching to using `async` - but Promise.all does handle non-promises:
```
> Promise.all([1, 2, 3, Promise.resolve(4)])
  .then((results) => results.forEach((r, i) => console.log(i,r)))
> Promise {
  <pending>,
  [Symbol(async_id_symbol)]: 612,
  [Symbol(trigger_async_id_symbol)]: 604
}

0 1
1 2
2 3
3 4
```

Reply:
Thats amazing. TIL.